src/static_allocator/device.rs

```diff
@@ -27,16 +24,117 @@ pub struct StaticDeviceAllocator {
+            self.0.insert(index, (size, backtrace));
+        }
+
+        fn remove(&mut self, index: &usize) {
+            self.0.remove(index);
+        }
+
+        fn block_count(&self) -> usize {
+            self.0.values().map(|&(size, _)| size).sum()
+        }
+
+        pub fn tail_index(&self) -> usize {
+            self.0
+                .last_key_value()
+                .map_or(0, |(&index, &(size, _))| index + size)
+        }
+
+        fn print(&self, detailed: bool, with_backtrace: bool) {
+            assert!(detailed || !with_backtrace);
+            if self.0.is_empty() {
+                println!("no allocations");
+                return;
+            }
+            println!("block_count: {}", self.block_count());
+            println!("tail_index: {}", self.tail_index());
+            const SEPARATOR: &str = "================================";
+            println!("{SEPARATOR}");
+            if detailed {
+                let mut last_index = 0;
+                for (index, (length, trace)) in &self.0 {
+                    let gap = index - last_index;
+                    last_index = index + length;
+                    if gap != 0 {
+                        println!("gap: {gap}");
+                        println!("{SEPARATOR}");
+                    }
+                    println!("index: {index}");
+                    println!("length: {length}");
+                    if with_backtrace {
+                        println!("backtrace: \n{trace}");
+                    }
+                    println!("{SEPARATOR}");
+                }
+            }
+        }
+    }
+
+    #[derive(Derivative)]
+    #[derivative(Clone, Debug, Default)]
+    pub struct AllocationStats {
+        pub allocations: Allocations,
+        pub allocations_at_maximum_block_count: Allocations,
+        pub allocations_at_maximum_block_count_at_maximum_tail_index: Allocations,
+    }
+
+    impl AllocationStats {
+        pub fn alloc(&mut self, index: usize, size: usize, backtrace: String) {
+            self.allocations.insert(index, size, backtrace);
+            let current_block_count = self.allocations.block_count();
+            let current_tail_index = self.allocations.tail_index();
+            let previous_maximum_block_count =
+                self.allocations_at_maximum_block_count.block_count();
+            if current_block_count > previous_maximum_block_count {
+                self.allocations_at_maximum_block_count = self.allocations.clone();
+            }
+            let previous_maximum_tail_index = self
+                .allocations_at_maximum_block_count_at_maximum_tail_index
+                .tail_index();
+            if current_tail_index > previous_maximum_tail_index {
+                self.allocations_at_maximum_block_count_at_maximum_tail_index =
+                    self.allocations.clone();
+            } else if current_tail_index == previous_maximum_tail_index {
```

Comment:
Nice, if there's a tie you prioritize stashing the spot with the higher block count.